[benchmark] Robust Measurements

palimondo opened the pull request:
With the faster benchmarks in Swift Benchmark Suite (now that the Legacy Factor refactoring is finished), we don’t need to be so extremely frugal with the sample count. Gathering just the first 3 to 10 samples per benchmark was not very representative from the statistical point of view. There was a not insignificant chance that the benchmark did not yet reach a steady state. I suspect it was the reason for sporadic false changes reported from CI. (For most recent example see #26455.)

This PR adjusts the measurement method (in both the `Benchmark_Driver` as well as `run_smoke_bench`) to sample each benchmark for 250 ms and gather at minimum 10 samples. The `run_smoke_bench` now gathers up to 10 independent runs for the suspected changes. Ten independent runs with larger sample counts are enough to get statistically sound results, so it no longer spins in a measurement loop waiting for 10 no-change runs in a row.

palimondo committed:
[benchmark] Fix parsing delta zeroed metadata

palimondo committed:
[benchmark] BenchmarkDriver min-samples & metadata

Support for invoking benchmark drivers with min-samples and gathering environmental metadata.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@eeckstein Hmm… I should have started full test, to also exercise the python unit tests. But if you trust me it all passes locally, we could check this (assuming smoke test passes) based on my honor word, to help Joe get less erratic results ASAP.

palimondo commented:
Hmm… I must have messed something up. Will check after returning from dog walk.

palimondo commented:
Interesting… failure in 2nd iteration of Onone in the `ObjectiveCBridgeASCIIStringFromFile` benchmark. @Catfish-Man?

palimondo commented:
@Catfish-Man, disregard. The error is in python code for merging results and has nothing with the particular benchmark.

Catfish-Man commented:
If we get more stable benchmark results can we decrease the 7% reporting threshold?

palimondo commented:
I can't figure out how am I sporadically reporting removed benchmarks whose runtime is 0 and how can they be missing in results dictionary during merge. Here it's too late now, I'll try again tomorrow. Sorry!

palimondo committed:
[benchmark] PerformanceTestResults merge samples

Adjusted how merged PerformanceTestResults track the number of underlying samples when using quantile subsampling.

palimondo commented:
This is some kind of a heisenbug… I'm going for vacation next week, so I will not be able to work on this. Maybe I'll push the little cleanup work I did in the meantime and see if that fixes the issue.

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci please test

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci please test

palimondo commented:
@swift-ci please test

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci please test

palimondo commented:
@swift-ci please test

palimondo commented:
This isn’t really better… 🙁

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
I’ve backed out the finer threshold. It was too high of a goal to get working correctly at this moment.

@atrick, @jckarter mentioned to me that you had your own analysis of the sporadic false reports, I’d appreciate if you could share your perspective here.

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
I think this looks OK, but let me run one more benchmark.

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci test

palimondo commented:
@swift-ci test

palimondo commented:
Well, benchmarks with runtimes under 20 μs need to be adjusted anyway… and `Flatten` has its issues and also a pending fix, but that PR needs a bit of a refresh: #20552… so I’d argue this is fine?

atrick commented:
Since my comments were alluded to above, I want to avoid confusion about the different kinds of variation we see in performance results. Performance is unpredictable in response to either changes to the compiler or to the inputs. I've made comments about these "optimizer instability" problems, or more generally performance unpredictability of the language. It's a problem because fundamentally good changes often produce bad results. The tradeoffs involved have to do with language features, peak benchmark performance, and time spent developing robust optimizations.

Run-to-run variation is different. It has to do with the design of the benchmarks and measurement methodology. I have not made any comments about this recently. The tradeoffs here are with turn around time, machine resources, precision and sensitivity, applicability to larger benchmarks, etc. To the extent that anyone pays attention to the benchmark scores, I do think it's important to fix "measurement instability". We just need to acknowledge that even with a perfectly deterministic benchmarking methodology, changes in benchmark scores across compiler changes can still be very misleading.

palimondo commented:
@eeckstein please review

eeckstein commented:
@palimondo I'll review this tomorrow. Please wait with merging until I get a chance to take a look.

palimondo commented:
Sure. I’m back from vacation, so I hope to get to a better solution for the ensuring more stable result than the last 3 commits (which reverted back to the 7% threshold and 16x loop), before you get to do the review.

eeckstein reviewed (changes requested):
FYI, when I tested the changes locally, I still saw some significant noise in some benchmarks.
This PR might improve the situation a bit, but it does not give us reliable "Robust Measurements" yet.

palimondo commented:
@eeckstein
> but it does not give us reliable "Robust Measurements" yet.

Agreed. This needs more work. I'll address your other comments inline.

palimondo commented on benchmark/scripts/run_smoke_bench:
```diff
@@ -141,26 +139,23 @@ def test_opt_levels(args):
     driver.tests = tests
-    return driver.run(num_samples=num_samples, sample_time=0.0025)
+    return driver.run(
+        num_iters=1, min_samples=10, sample_time=0.25, quantile=20,
```
It is not a typo. I've intended to gather samples for a quarter of a second to ensure benchmarks have reached steady state. From PR description:
> This PR adjusts the measurement method (in both the Benchmark_Driver as well as run_smoke_bench) to sample each benchmark for 250 ms and gather at minimum 10 samples. The run_smoke_bench now gathers up to 10 independent runs for the suspected changes. Ten independent runs with larger sample counts are enough to get statistically sound results, so it no longer spins in a measurement loop waiting for 10 no-change runs in a row.

Of course it makes the measurement take longer than before, because we are gathering more samples. The 2500 μs you've used originally were a reasonable trade-off at time when measuring the whole benchmark suite took much longer. But the proper baseline to compare this is before all the **legacy factor** refactoring. These have massively shortened the time to run the whole SBS. The point here is to use the time we've won and invest it into better quality of measurement: more samples. This way we can rely on better statistical methods in our analysis.

Here are my local timings per one full run of SBS with in various cases:

options | `O` | `Osize` | `Onone`
--|--|--|--
`--sample-time=0.0025 --num-samples=3` | 0:10 | 0:12 | 1:31
`--sample-time=0.0025 --num-samples=10` | 0:31 | 0:39 | 4:43
`--sample-time=0.25 --min-samples=10`<br>` --num-iters=1` | 1:31 | 2:07 | 6:12

There were 758 benchmarks in the pre-commit category of the SBS when I measured this locally, which — at 0.25 s per benchmark — should nominally take **3:10** for one full run of SBS.  Comparing this to the last line with my proposed measurement parameters, `O` and `Osize` take less time, because measurements are capped at 200 samples and end earlier for fast benchmarks. The `Onone` takes more than that because on my slow machine it takes longer to gather at least 10 samples per benchmark (that's also visible in the `--num-samples=10` case, which be the worst case scenario per iteration in our current system).

Your modern machines are much faster. Looking at the [console of the benchmark run](https://ci.swift.org/job/swift-PR-osx-perf/1156/console) in this PR and comparing it to [ numbers from another benchmark run](https://ci.swift.org/job/swift-PR-osx-smoke-perf/924/console) of the current `run_smoke_bench` from PR 26630 yields following times for single 0 iteration of SBS:

options | `O` | `Osize` | `Onone`
--|--|--|--
`--sample-time=0.0025 --num-samples=3` | 0:06 | 0:07 | 0:31
`--sample-time=0.25 --min-samples=10`<br>` --num-iters=1` | 0:34 | 0:39 | 2:21

The total time to run smoke benchmark was 3:19 with current method and 8:09 with the one in this PR, a slowdown of approximately 2.5. Note that the time from initial git pull to finished compilation on this PR was 19:23.

palimondo replied:
Note that going back to capping the iterations at 10 (after addressing the “instabilities”), might slightly improve the total runtime of this method, to maybe somewhere around 2x slowdown.

palimondo replied:
I will also look at running benchmarks in parallel with this method, which will speed this back up again, depending on the number of cores on the machine.

palimondo commented on benchmark/scripts/Benchmark_Driver:
```diff
@@ -183,8 +183,9 @@ class BenchmarkDriver(object):
-                      [self.run(test, measure_memory=True,
-                                num_iters=1, quantile=20)
+                      [self.run(test, num_iters=1, min_samples=10,
+                                sample_time=0.25, quantile=20,
```
This is the original `Benchmark_Driver`, which used default sample time of 1 second up until now. This is a 4x speedup of its measurement throughput. From the commit description in b4c8153:
> With the faster benchmarks in the Swift Benchmark Suite, we now don’t need to spend a whole 1 second measuring each one of them. So I’m adjusting the Benchmark_Driver to sample each one for 250 ms. Quarter of a second should be plenty of time to reach a stable state in most cases.
>
> For the slow ones and measuring of the unoptimized Onone build, let’s improve the quality of statistics by gathering at least 10 samples.

palimondo commented on benchmark/scripts/run_smoke_bench:
```diff
@@ -170,35 +165,27 @@ def merge(results, other_results):
+    changed = tests.decreased + tests.increased + tests.added
 
-    while len(changed) > 0 and unchanged_length_count < 10:
+    while len(changed) > 0 and i < 16:
```
The last 3 commits here were done from mobile and were my attempt to somehow address the apparent instability by reverting to the old brute force methods. I'm pretty sure that my original approach to stop at 10 measurements was a correct one, but I need to handle the special cases of benchmarks with runtimes under 20 μs as a special case in result comparison and also adding filtering for min outliers like is the case with `ObjectiveCBridgeStubToNSDateRef`.

palimondo replied:
![IMG_0747](https://user-images.githubusercontent.com/1046899/62961273-c077dd80-bdfc-11e9-90b6-d0bd6e764bfe.jpeg)

palimondo replied:
Another example of benchmark that jumped up here during the first run, which seems like something that has about 2:10 chance to look unstable is we stick to the naive comparison of MIN values is `DataAppenDataSmallToSmall`: 
![IMG_0745](https://user-images.githubusercontent.com/1046899/62961580-527fe600-bdfd-11e9-9e8a-899855794ff5.jpeg)

palimondo replied:
I’d like to fix that issue by expanding on the dubious indicator **(?)** in the ratio column, so that it revealed more details about the underlying sample distribution, to enable us to make more informed decisions in such cases.

palimondo committed:
[benchmark] Don’t justify last column in reports

In text reports, don’t justify the last columns with unnecessary spaces.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
This is still work in progress. I have modified the benchmark reporting for dubious result comparisons to print out empirical sample distribution ([ventiles](https://en.wiktionary.org/wiki/ventile)), to get better insight into the nature of problematic benchmarks. In the current process, the reported changes have samples gathered from 10 independent measurements. These are merged together the form a composite sample distribution.

In my local measurements, there is a class of benchmarks that tend to show up in the reports as false changes, given that I am comparing the same binary against itself. I'm interested to see to what extent are these HW specific and to what extent are these _"unstable"_ benchmarks platform independent.

Next step would be to devise a heuristic, other than comparing just MIN values between NEW and OLD, that deals with the perceived instability caused by this naive approach.

palimondo commented on benchmark/scripts/test_compare_perf_tests.py:
```diff
@@ -507,6 +507,13 @@ class OldAndNewLog(unittest.TestCase):
                              for line in
                              new_log_content.splitlines()])])
 
+    old_results['D'] = PerformanceTestResult(
```
This is a visualization of an example of the problematic benchmark used in these unit tests:
![Dubious Benchmark](https://user-images.githubusercontent.com/1046899/64974128-86db3c00-d8ac-11e9-8b44-1593e5b24e97.png)

palimondo replied:
These are two independent executions, picked to showcase the rarely occurring but very significant variation present in one of them.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
There are clearly a few benchmarks with too small runtime, some of them because of wrong `legacyFactor`s due to the large performance gap between my machine and the CI machines, so I’m running a full `BenchmarkDoctor` check on CI to get a list of those that need to be re-adjusted.

palimondo commented:
Hah, I had to pull the benchmark check from the console log, because I’ve missed a conditional to print it in HTML. But I got that list, don’t worry.

palimondo committed:
[benchmark] Retire old num-iters in verbose format

After commit 331c0bf772a4626aa1f8e1f8aae7629e10d350db from a year ago, all samples from the same run have the same num-iters.

palimondo committed:
[benchmark] [Gardening] Fix assertEqual naming

palimondo committed:
[benchmark] Add tests for all_samples

palimondo committed:
[benchmark] Refactor: store num_iters on PTS

Store the number of iterations averaged in each sample on the PerformanceTestSamples.

palimondo committed:
[benchmark] Refactor: remove class Sample

Removed Sample class, that was previously holding num_iters and the ordinal number of the sample.

palimondo committed:
[benchmark] Refactor: simpler exclude_outliers

palimondo committed:
[benchmark] More immutable PerformanceTestSamples

Removed the ability to add individual samples to PTS.
PerformanceTestSamples are technically not really immutable, because of exclude_outliers.

palimondo commented:
The purpose of this PR is to find more statistically rigorous method of detecting performance changes in our benchmarks, while minimizing [type I and type II errors](https://en.wikipedia.org/wiki/Type_I_and_type_II_errors). The altered measurement technique has apparently increased the chance of type I errors — false positives — because it has eliminated the averaging of multiple samples and has significantly increased the number of samples gathered.

From the ventiles collected from the altered dubious reports (they are marked by **(?)** in the Ratio column) I’ve charted the runtimes to see what is the empirical distribution of the samples from benchmarks that are popping up with this more sensitive measurement technique.

## Flatten
Most notorious are the Flatten benchmarks, which sporadically appear even with the current measurement process:
![IMG_0874](https://user-images.githubusercontent.com/1046899/68936065-02b51000-079a-11ea-82f5-3c537706a906.jpeg)
![IMG_0875](https://user-images.githubusercontent.com/1046899/68936075-0f396880-079a-11ea-9b80-e4d19a6bd9b4.jpeg)

These clearly demonstrate that for certain benchmarks, using the [minimum as a location estimator is genuinely problematic](https://palimondo.github.io/robust-microbench/detecting-changes.html#issues-with-minimum-as-location-estimator).

We have to remember that the samples reported in tables and visualized in the charts above are aggregates from 10 independent measurement runs. But here we don’t have all the samples, just the minimum and [ventiles](https://en.wiktionary.org/wiki/ventile#English), so the lower end of the distribution is also over-represented (10x MIN and 5th percentiles). It is not clear if the extreme outliers appear rarely in every measurement or if they come from one measurement with different distribution.

We should probably fix these benchmarks by lowering their runtime anyway (#20552), but they present interesting challenge, which we could try to solve by applying more sophisticated statistical techniques, than just looking if there’s more than 5% difference between the minimum values.

## ObjectiveCBridgeStubToNSDateRef
This benchmarks has slightly different empirical distribution between the `O` and `Onone`:
![IMG_0876](https://user-images.githubusercontent.com/1046899/68937998-bbc91980-079d-11ea-9f7d-e54faa8ecea7.jpeg)
![IMG_0877](https://user-images.githubusercontent.com/1046899/68938012-c4215480-079d-11ea-936d-522df3aae708.jpeg)

Here again, minimum proves to be poor location estimate for this benchmark, because it appears to have the highest variance, when compared to other ventiles. Interestingly, the sample distribution of this benchmarks is also quite wide, with [interquartile range](https://en.wikipedia.org/wiki/Interquartile_range) 12-14%.

## Others
Most other falsely reported benchmarks show that the minimum can be noisy location indicator, compared to the 5th or 10th percentile.

(To be continued...)

palimondo committed:
[benchmark] Fix crash in oversampled results

Gracefully handle the parsing of oversampled values in critical configuration, when the sampling error causes an ommision of certain quantiles from the report.

futurejones commented:
Can one of the admins verify this patch?

palimondo commented:
> Can one of the admins verify this patch?

I’m still working on this. Just got bogged down in a job that actually pays the bills… 😐

palimondo committed:
[benchmark] Report ventiles for dubious results

For dubious result comparisons, print out empirical sample distribution (ventiles) to enable humans to reach informed decisions about these performance changes.

palimondo committed:
[benchmark] Refactor: remove redundant _runtimes

palimondo committed:
[benchmark] Collate metadata when merging PTRs

Handle optional `--meta` data in `merge` of `PerformanceTestResults`. Pick minimum from memory pages, sum the number of involuntrary contex switches and yield counts.

palimondo committed:
[benchmark] Keep merged independent run samples

When merging `PerformanceTestResults`s keep the original `PerformanceTestSample`s from all independent runs. These will be used to choose the most stable (least variable) location estimate for the `ResultComparison` down the road.

palimondo commented:
My current plan to make this measurement method more robust:
1. When merging `PerformanceTestResult`s, keep the original `PerformanceTestSample`s from all independent runs. 
2. These will be used to choose the most stable (least variable) location estimate for the `ResultComparison`: compute variance for MIN, V1, V2, Q1, MED (these are 0th, 5th, 10th, 25th a 50th percentiles respectively) for OLD and NEW result (these will be composed from at least 3 and up to 10 independent runs in the `run_smoke_benchmark`), multiply these<sup>*</sup> with each other and pick the one with lowest value as the most stable location estimate for the sample distributions. This will then be used in the comparison as usual.
3. Adjust the `is_dubious` analysis on `ResultComparison` to be a bit more sophisticated than the current range overlap between the OLD and NEW samples. Keep the ventiles printout that shows the cumulative empirical sample distribution to aid in decision making should such case arise.

<sup>*</sup> *Since variance is standard deviation squared, using a simple multiplication as formula to pick the most stable location should work fine, thanks to its exponential nature… But I’ll try to simulate and visualize this with few plots in R tomorrow, before implementing this strategy in Python.*

CC @eeckstein @gottesmm 
Please don’t push forward with #30085 yet, temporarily reverting #29719 to land this is complicated enough. I have sunk a lot of thought and work into this PR and think it will be a major improvement to our measurement methodology. I’ll finish this over the weekend and will help with the Python 3 conversion thereafter.

gottesmm commented:
@palimondo if you are going to do this, my suggestion is to do this in a way that you can land this without flipping the final switch. That way we do not rush this in and all of a sudden @swift-ci benchmark changes all of a sudden in a rushed way. I am concerned with putting things into the tree just to get them into the tree.

By landing it we can make sure that the current approach still works, we can flip the switch gracefully when you have time to fully look into any problems that come up. I am assuming since you said that you need to use time to work your normal job that it may be difficult for you to spend more time immediately after landing this if any issues come up. Is my interpretation correct?

palimondo committed:
[benchmark] Override all_samples on merged results

To save on memory used by merged `PerformanceTestResult`s, the rarely used `PerformanceTestSample.all_samples` can gather the samples on demand from the result’s `independent_runs` instead of keeping another copy.

palimondo committed:
[benchmark] Python 3 Support

palimondo committed:
Revert "[NFC][Python: black] Reformatted the benchmark Python sources using utils/python_format.py."

palimondo commented:
@gottesmm 
> By landing it we can make sure that the current approach still works, we can flip the switch gracefully when you have time to fully look into any problems that come up. I am assuming since you said that you need to use time to work your normal job that it may be difficult for you to spend more time immediately after landing this if any issues come up. Is my interpretation correct?

I’ll keep the final “flipping the switch" as last few commits, so that we can separate them as necessary. But @eeckstein’s reviews and merges usually happen in my after work hours anyway, so my evenings are all yours if the need to fix anything will arise. (The bill paying job was just a reference to my dashed dreams of turning the work on Swift into paying remote job. Apple’s hiring policy that requires relocation even for work on distributed open source project is just frustrating… 😞)

palimondo committed:
[benchmark] Dubious indicator for changes only

Display the `(?)` indicator for dubious result only for changes, never for the unchanged results.
Refactored `ResultComparison.init` with simplyfied range check.

gottesmm commented:
@palimondo thanks. As long as we can revert easily flipping the switch, I am 100% ok! Make sure to separate your changes out carefully!

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
OK, there is probably still some fine tuning work ahead of us, but let’s see if the basic idea is correct. I’ll run a few benchmarks to gather more data.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
This is a bit unexpected, but I see this also in my local testing. Let me dump quartiles for all independent runs at once and I’ll do some analysis tomorrow.

eeckstein commented:
@palimondo To reiterate: increasing the sample time in run _smoke_bench by 100x is not acceptable. I cannot approve this PR with this change.

palimondo commented:
@eeckstein I understand your point. Let’s make this more manageable:
* I have split off the groundwork for robust measurements to it own PR #30158. It was the bulk of the work here and the part I really don’t want to loose, because of all the conflicts coming from automatic formatting changes from #29719. I want to land this ASAP.
* We can discuss the measurement parameters and the merits of proposed changes from this PR later. It will be much smaller change once the groundwork lands and it doesn’t have to block the Python 3 support and all the automatic formatting changes. 
* I’d like to hear more of your thoughts about that, since #29719 has landed without any justification in its description. Was there some plan discussed in Swift forums that I missed? I’m assuming @Rostepher’s work on that will can be reproduced by some automated step (missing docs?)?

palimondo commented:
@swift-ci please benchmark

Rostepher commented:
@palimondo The goal is to add another Python validation-test to run `utils/python_format.py --check` which should enforce the black style on all our Python sources. Details are in #29701.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo committed:
[Gardening] Refactored runtimes correction for SO

Replaced list comprehension that computes the minimum from runtimes corrected for setup overhead with a procedural style that is easier to understand.

gottesmm commented:
I don't understand this new output and it definitely needs to be behind a detail drop down.

palimondo commented:
@gottesmm I’m sorry for spamming you all with these verbose reports. I needed these just temporarily to gather more data and devise better strategy for coping with the issue of false changes. Now it’s clear the extreme outliers are coming from “lucky runs”, which then skew the aggregated sample distribution, used for the final comparison of OLD and NEW. The selection of most stable location from f888aefe9d989fee26fac6a2f8ef3d6deeb6c15c does not help with this problem, because the individual runs have the same shape of distribution, the lucky runs are just shifted in location and we keep comparing the MINs. I think I’ll need to detect and throw out such outlier runs. This explains why `run_smoke_bench` currently needs to go up to 20 or more iterations to get the “lucky run” from the other result and doesn’t always succeed.

palimondo committed:
[benchmark] Compare with stable location estimate

Use the most stable location estimate (MIN, P05, P10, Q1, MED) for `ResultComparison` based on the standard deviation for given quantiles across the independent runs and the aggregate sample (hardening against outlier runs). We multiply the standard deviations for each quantile between the OLD and NEW results to get “cross-sample” variance. We pick the quantile with lowest variance as most stable and use it for the comparisons.

palimondo committed:
[benchmark] Benchmark_Driver: faster iterations

With the faster benchmarks in the Swift Benchmark Suite, we now don’t need to spend a whole 1 second measuring each one of them. So I’m adjusting the Benchmark_Driver to sample each one for 50 ms and gather up to 102 actual runtime values (percentiles + MIN + MAX). In my tests, for most optmized benchmarks, the sample distribution was still roughly comparable with full second mesurements. It is more important to gather samples from multiple independent runs to cover the possible variations in distribution.

palimondo committed:
[benchmark] run_smoke_bench: more robust sampling

With the faster benchmarks in Swift Benchmark Suite (now that the Legacy Factor refactoring is finished), we don’t need to be so extremly frugal with the sample count. Gathering just the first 3 to 10 samples per benchmark was not very representative from the statistical point of view. I suspect it hides Type II errors — unreported changes.

Adjusting the measurement method to sample each benchmark for 50 ms and gather at minimum 10 samples. For the suspected changes, gather up to 10 independent samples.

Also thorougly measure the newly added test in re-runs.

palimondo committed:
[benchmark] run_smoke_bench cleanup

Removed unused `num_samples` argument and redundant `run_count` variable.
Also gather metadata (currently unused).
Print the actual threshold in the “How to read the data“ decription.

palimondo commented:
@swift-ci please python lint

palimondo commented:
@swift-ci please python lint

palimondo commented:
@swift-ci python lint

palimondo commented:
@swift-ci python lint

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
@swift-ci benchmark

palimondo commented:
Wow, now this is upsetting! Switching away from MIN as location estimate on these benchmarks did backfire tremendously. I have seen nothing like this locally. I need to re-enable that verbose dump to understand the distribution of individual runs for these cases.

palimondo committed:
[DNM] Dump quartiles for all independent runs

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@shahmishal @eeckstein Did @swift-ci start to invoke `run_smoke_bench` with `--num-reruns` equal to 22 recently? I’m surprised by the number of independent iterations I’ve dumped. It seems we were using the default value just few days ago…

CodaFi commented:
@palimondo This PR is enormous! The python formatting changes need to be broken out separately from the benchmark formatting changes.

palimondo commented:
@CodaFi Python formatting was broken out already in a separate PR this one
depends on, but that's neither here nor there since #30085 has landed :-/
resolving the merge conflicts will be a slog. I'll get back to this,
eventually.

On Wed, Apr 15, 2020 at 2:17 AM Robert Widmann <notifications@github.com>
wrote:

> @palimondo <https://github.com/palimondo> This PR is enormous! The python
> formatting changes need to be broken out separately from the benchmark
> formatting changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/26462#issuecomment-613744619>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH7S4YDI4HRXNKCVDDZBTDRMT4KDANCNFSM4II66NHQ>
> .
>

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
